src/mon/OSDMonitor.cc

```diff
@@ -6890,30 +6917,38 @@ int OSDMonitor::prepare_command_osd_new(
     return EEXIST;
   }
 
+  string device_class;
+  auto p = params.find("crush_device_class");
+  if (p != params.end()) {
+    device_class = p->second;
+    dout(20) << __func__ << " device_class will be " << device_class << dendl;
+  }
   string cephx_secret, lockbox_secret, dmcrypt_key;
   bool has_lockbox = false;
-  bool has_secrets = (!secrets.empty());
+  bool has_secrets = params.count("cephx_secret")
```

Comment:
this breaks the tests at https://github.com/ceph/ceph/blob/master/qa/workunits/cephtool/test.sh#L1269 . before this change, before this change the `secrets` is a map of `{"asd" => "", "}" => ""}`, so it's not "empty".  and a non-empty secret map without "cephx_secret" is a considered invalid input before this change.

but with this change, we practically ignore the "params" if it does not contain anything interesting. even if it is a map parsed from gibberish.

so my question is: shall we validate the params by checking if it contains some keys we do not understand? something like
```c++
for (auto param& : params) {
  if (param.first == "cephx_secret")) {
    has_secrets = true;
  } else if (param.first == "cephx_lockbox_secret") {
    has_secrets = true;
    has_lockbox_secret = true;
  } else if (param.first == "dmcrypt_key") {
    has_secrets = true;
    has_dmcrypt_key = true;
  } else if (param.first == "crush_device_class") {
    device_class = param.second;
  } else {
    ss << "unknown param " << param;
    return -EINVAL;
  }
}
```